src/Logic.hs

```diff
@@ -634,25 +637,66 @@ unintegrateAfter pid state = case Pr.lookupPullRequest pid state of
 
 -- | If there is an integration candidate, and its integration sha matches that of the build,
 --   then update the build status for that pull request. Otherwise do nothing.
-handleBuildStatusChanged :: Sha -> BuildStatus -> ProjectState -> Action ProjectState
-handleBuildStatusChanged buildSha newStatus state = pure $
-  compose [ Pr.updatePullRequest pid setBuildStatus
-          . case newStatus of
-            BuildFailed _ -> unintegrateAfter pid
-            _             -> id
+handleBuildStatusChanged :: Sha -> Context -> BuildStatus -> ProjectState -> Action ProjectState
+handleBuildStatusChanged buildSha check newStatus state = pure $
+  compose [ Pr.updatePullRequest pid (setCheckIfNeeded . setBuildStatus)
+          . unintegratePullRequest pid
```

Comment:
Nit: This sounds like it's unconditionally unintegrating the PR, which is slightly confusing.